### PR TITLE
Nwb refactor metadata

### LINKED ISSUE #320: error converting from OpenEphys to NWB
It looks like there is an issue converting from OpenEphys to NWB when location information is missing:

```python
from spikeextractors import OpenEphysRecordingExtractor, NwbRecordingExtractor

fpath = '/Volumes/GoogleDrive/Shared drives/ePhys_Processing_Soltesz/JF_298_003_2020-02-04_15-15-40'
re = OpenEphysRecordingExtractor(fpath)
out_path = '/Volumes/GoogleDrive/My Drive/data/Soltesz/test.nwb'
NwbRecordingExtractor.write_recording(re, out_path)
```

```
Reading all channels
100_CH1.continuous
100_CH2.continuous
100_CH3.continuous
100_CH4.continuous
100_CH5.continuous
100_CH6.continuous
100_CH7.continuous
100_CH8.continuous
100_CH9.continuous
100_CH10.continuous
100_CH11.continuous
100_CH12.continuous
100_CH13.continuous
100_CH14.continuous
100_CH15.continuous
100_CH16.continuous
100_CH17.continuous
100_CH18.continuous
100_CH19.continuous
100_CH20.continuous
100_CH21.continuous
100_CH22.continuous
100_CH23.continuous
100_CH24.continuous
100_CH25.continuous
100_CH26.continuous
100_CH27.continuous
100_CH28.continuous
100_CH29.continuous
100_CH30.continuous
100_CH31.continuous
100_CH32.continuous


/Users/bendichter/opt/anaconda3/lib/python3.7/site-packages/pynwb/file.py:694: UserWarning: Date is missing timezone information. Updating to local timezone.
  warn("Date is missing timezone information. Updating to local timezone.")
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-8-5bba4172f62d> in <module>
      1 out_path = '/Volumes/GoogleDrive/My Drive/data/Soltesz/test.nwb'
----> 2 NwbRecordingExtractor.write_recording(re, out_path)

~/opt/anaconda3/lib/python3.7/site-packages/spikeextractors/extractors/nwbextractors/nwbextractors.py in write_recording(recording, save_path, acquisition_name, **nwbfile_kwargs)
    320             for m in channel_ids:
    321                 if m not in nwb_elec_ids:
--> 322                     location = recording.get_channel_property(m, 'location')
    323                     while len(location) < 2:
    324                         location = np.append(location, [0])

~/opt/anaconda3/lib/python3.7/site-packages/spikeextractors/recordingextractor.py in get_channel_property(self, channel_id, property_name)
    393             raise ValueError('no properties found for channel' + str(channel_id))
    394         if property_name not in self._channel_properties[channel_id]:
--> 395             raise RuntimeError(str(property_name) + " has not been added to channel " + str(channel_id))
    396         if not isinstance(property_name, str):
    397             raise TypeError(str(property_name) + " must be a string")

RuntimeError: location has not been added to channel 0
```

Similar error when requesting location from the `OpenEphysRecordingExtractor` object:

```python
re.get_channel_locations()
```

```python
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-11-455d3dd614d7> in <module>
----> 1 re.get_channel_locations()

~/opt/anaconda3/lib/python3.7/site-packages/spikeextractors/recordingextractor.py in get_channel_locations(self, channel_ids)
    244         locations = []
    245         for channel_id in channel_ids:
--> 246             location = self.get_channel_property(channel_id, 'location')
    247             locations.append(location)
    248         return locations

~/opt/anaconda3/lib/python3.7/site-packages/spikeextractors/recordingextractor.py in get_channel_property(self, channel_id, property_name)
    393             raise ValueError('no properties found for channel' + str(channel_id))
    394         if property_name not in self._channel_properties[channel_id]:
--> 395             raise RuntimeError(str(property_name) + " has not been added to channel " + str(channel_id))
    396         if not isinstance(property_name, str):
    397             raise TypeError(str(property_name) + " must be a string")

RuntimeError: location has not been added to channel 0
```

### PR DESCRIPTION
**Pull request by luiztauffer**
This refactor contains some changes that help us integrate NwbRecordingExtractor better with many of our current projects, plus some fixes:

- add metadata control to NwbExtractor. This allows the use of metadata file instructions for NWB file contruction, a practice we’ve been trying to promote. It is optional and the standard conversions recording <--> nwb should work just the same without it.

- breaks down the nwb file formatting code into smaller, nwb-class-specific static methods, which allows us to interact better with each step of the way if needed for more advanced cases. Again, it won’t alter the standard conversions recording <--> nwb.

- fix #320

- fix temporarily #323

### CODE REVIEW

File: spikeextractors/extractors/nwbextractors/nwbextractors.py [hunk 1]
```diff
@@ -118,6 +118,15 @@ def most_relevant_ch(traces):
     return relevant_ch
 
 
+def update_dict(d, u):
```
Comment (bendichter) [inline spikeextractors/extractors/nwbextractors/nwbextractors.py]: What is the purpose of this? Why can't you use `dict.update`?

File: spikeextractors/sortingextractor.py [hunk 2]
```diff
@@ -18,7 +18,7 @@ def __init__(self):
         self._unit_properties = {}
         self._unit_features = {}
         self._sampling_frequency = None
-        self.id = np.random.randint(low=0, high=9223372036854775807)
+        self.id = np.random.randint(low=0, high=9223372036854775807, dtype='int64')
```
Comment (bendichter) [inline spikeextractors/sortingextractor.py]: why the change? What was the issue with this being an `int`?

File: spikeextractors/extractors/nwbextractors/nwbextractors.py [hunk 3]
```diff
@@ -204,6 +220,36 @@ def __init__(self, path, electrical_series_name='ElectricalSeries'):
...
+                    'location': v.location,
+                    'device': v.device.name
+                })
+            # Update metadata with ElectricalSeries info
+            self.nwb_metadata['Ecephys']['ElectricalSeries'] = []
+            self.nwb_metadata['Ecephys']['ElectricalSeries'].append({
+                'name': es.name,
+                'description': es.description
+            })
+
```
Comment (bendichter) [inline spikeextractors/extractors/nwbextractors/nwbextractors.py]: I would make this section its own static function

**Comment by colehurwitz** [comment]
@bendichter @luiztauffer is this still a WIP or are you happy with the current state of the NWB extractors?

**Comment by bendichter** [comment]
@colehurwitz Thanks for the nudge. The importer should work as it currently stands for accessing tools like the spike sorters and visualization tools. It currently writes valid NWB files, but does not have the flexibility to provide the user the opportunity to populate optional meta-data fields that may be essential information for sharing the data. Since we had met our original goals, and due to the dynamic nature of our different projects, this line of development has been on the back burner for a little while we reach goals with our other ongoing projects. If you don't mind, I'd like to keep this open so that we can continue to move toward helping NWB users conveniently package their data with all of the essential meta-data necessary for sharing.

**Comment by alejoe91** [comment]
@luiztauffer note that I force pynwb version in the [.travis](https://github.com/SpikeInterface/spikeextractors/blob/master/.travis.yml#L9). You can update to the newest version

File: spikeextractors/extractors/nwbextractors/nwbextractors.py (hunk 1 above)
Comment (luiztauffer) [reply spikeextractors/extractors/nwbextractors/nwbextractors.py]: See #303

File: spikeextractors/sortingextractor.py (hunk 2 above)
Comment (luiztauffer) [reply spikeextractors/sortingextractor.py]: see #303

File: spikeextractors/extractors/nwbextractors/nwbextractors.py (hunk 1 above)
Comment (luiztauffer) [reply spikeextractors/extractors/nwbextractors/nwbextractors.py]: this does a "deep updating"

**Comment by luiztauffer** [comment]
@alejoe91 ok I will!

File: spikeextractors/extractors/nwbextractors/nwbextractors.py (hunk 3 above)
Comment (luiztauffer) [reply spikeextractors/extractors/nwbextractors/nwbextractors.py]: done on 1242f2c

**Comment by bendichter** [comment]
These changes look good to me!

**Comment by alejoe91** [comment]
Looks good guys! :) ready to merge? I suggest we make a section in the spikeinterface docs on how to export properly to NWB later

**Comment by bendichter** [comment]
@alejoe91 yes, ready when you are :-)
